### PR TITLE
Replace TryFrom with custom TryFromJsValue trait

### LINKED ISSUE #3685: 0.2.88 implements TryFrom on all ABI-bound objects, breaking API bindings that have custom TryFrom implementations
`0.2.88` automatically derives `TryFrom` trait for any struct that uses `#[wasm_bindgen]`.

This prevents any developer from creating custom `TryFrom<JsValue>` conversions for the ABI-bound struct.

We have a rather large Rust/JS API that uses `TryFrom` to provide custom processing for JS objects. For example, JS code can pass a Rust struct, a buffer, or a hex string, which the `TryFrom` resolves. This is extremely useful because JS objects can be of different types. The object received from JS does not need to be an ABI-bound object in order to be converted to a native Rust struct.

This addition makes it no longer possible to create custom `TryFrom` handling between `JsValue` and a Rust struct that is also WASM ABI-bound.

I believe that such imposition should not be present in the attribute macro, or it should be possibly gated by a feature or a macro attribute. It does not seem quite right that an attribute macro creates functionality that is typically created by derive macros.  

This is semantically not different than (for example) `wasm_bindgen` implementing a custom `Clone` trait, preventing the developer from being able to implement custom `Clone` functionality for his objects...

Perhaps there should be a `FromJsValue` derive macro available?  Or such conversion should be moved into a custom `try_from_js_abi()` function implemented on each ABI-bound object? (implementing a function will create much less of a potential interference with underlying frameworks).

### PR DESCRIPTION
**Pull request by biryukovmaxim**
This commit replaces the usage of the standard library's TryFrom trait with a custom TryFromJsValue trait added for handling the conversion from JavaScript values in WASM. This trait has been implemented for different types across multiple modules to ensure consistent error handling on failed conversions.

fixes #3685

### CODE REVIEW

**Review by daxpedda — CHANGES_REQUESTED** [review]
Could you add an entry to the changelog as well?
LGTM otherwise.

**Comment by daxpedda** [comment]
@hamza1311 and @Liamolucko, the idea here is to yank the last version and a new one with this change.
Myself I'm not entirely fond of the whole idea, but I believe if somebody is affected and willing to provide a fix as well, we should take that as something that needs to be addressed. So I would vote in favor.

Let me know what you think.

**Comment by Liamolucko** [comment]
> Myself I'm not entirely fond of the whole idea, but I believe if somebody is affected and willing to provide a fix as well, we should take that as something that needs to be addressed. So I would vote in favor.

I feel the same way.

This does un-fix #2231 though, since `wasm_bindgen::convert` is an internal module and so `TryFromJsValue` is effectively private.

The easy solution is just to move it to `wasm_bindgen::TryFromJsValue`, but it's a bit weird to make people use a custom trait rather than the standard `TryFrom<JsValue>`. Alternatively we could make a `TryFromJsValue` derive macro, but that's a lot more work.

**Comment by biryukovmaxim** [comment]
> This does un-fix #2231 though, since `wasm_bindgen::convert` is an internal module and so `TryFromJsValue` is effectively private.

convert module is public, it already exports From/IntoWasmAbi, and others. But these two looks like custom `From` and `Into` that include additional constraints.

**Comment by Liamolucko** [comment]
> convert module is public, it already exports From/IntoWasmAbi, and others. But these two looks like custom `From` and `Into` that include additional constraints.

Yes, it's _technically_ public, but see its documentation:

> This is mostly an internal module, no stability guarantees are provided. Use at your own risk.

It's only really public so that macros can access it, nobody be using anything from in there directly.

**Comment by biryukovmaxim** [comment]
> > This is mostly an internal module, no stability guarantees are provided. Use at your own risk.
> 
> It's only really public so that macros can access it, nobody should be using anything from in there directly.

Oh. I see. Your suggestion is to move the trait to wasm_bindgen - root module?

**Comment by biryukovmaxim** [comment]
> Could you add an entry to the changelog as well? LGTM otherwise.

done

**Comment by daxpedda** [comment]
> Alternatively we could make a `TryFromJsValue` derive macro, but that's a lot more work.

Even if we want to do that one day, it would still require `TryFromJsValue` trait to be public.
I agree though this is weird, but I can't come up with a better solution.

File: src/lib.rs
Comment (daxpedda) [inline src/lib.rs:1916]: This will require some documentation now.

File: CHANGELOG.md
Comment (daxpedda) [inline CHANGELOG.md:16]:
```suggestion

* Fixed an issue where `#[wasm_bindgen]` automatically derived the `TryFrom` trait for any struct, preventing custom `TryFrom<JsValue>` implementations. It has been updated to utilize a new `TryFromJsValue` trait instead.
  [#3709](https://github.com/rustwasm/wasm-bindgen/pull/3709)
```

**Review by daxpedda — CHANGES_REQUESTED** [review]

**Comment by daxpedda** [comment]
I'm on the fence about moving `TryFromJsValue` just back to `convert` and just figure out what to do here another time.
Any input would be appreciated.

File: src/convert/traits.rs
Comment (daxpedda) [inline src/convert/traits.rs:266]:
```suggestion
/// [`TryFromJsValue`] is a trait for converting a JavaScript value ([`JsValue`])
/// into a Rust type. It is used by the [`wasm_bindgen`](wasm_bindgen_macro::wasm_bindgen)
/// proc-macro to allow conversion to user types.
///
/// Types implementing this trait must specify their conversion logic from
/// [`JsValue`] to the Rust type, handling any potential errors that may occur
/// during the conversion process.
```

I removed the historical part, which I would like to pretend just never happened instead of having to remember to remove it at some later point when it becomes only confusing instead of useful for new users.

Also adjusted some of the wording implying that this is anything else then internal stuff for now.

**Review by daxpedda — CHANGES_REQUESTED** [review]
LGTM, just some nits about docs.

As mentioned before, this unfixes #2231, but we can figure out exactly what to do here at some later point.

@biryukovmaxim you should probably hold off on any major changes for now, we are still discussing how exactly to proceed, there's no need to move `TryFromJsValue` back and forth upon every whimsical comment I make! (don't move it back now :smile:)

**Comment by dani-garcia** [comment]
I've hit a problem with the TryFrom changes that is unrelated to #3685, but that would probably also occur with this PR, related to the ambiguity of using `Self::Error` when `#[wasm_bindgen]` is applied to an enum containing an `Error` variant.

```
#[wasm_bindgen]
pub enum LogLevel {
    Trace,
    Debug,
    Info,
    Warn,
    Error,
}
```

The way to solve that would be to use the fully qualified syntax for `Self`:
```diff
fn try_from_js_value(value: #wasm_bindgen::JsValue)
-                   -> #wasm_bindgen::__rt::std::result::Result<Self, Self::Error> {
+                   -> #wasm_bindgen::__rt::std::result::Result<Self, <Self as TryFromJsValue>::Error> {
```

I can open a PR myself after this one is merged if you prefer to merge this one as-is though.

**Comment by daxpedda** [comment]
Could you open a PR based targeting `main`? We can merge that right now and let this PR rebase.
Thanks for bring this up!

**Comment by Liamolucko** [comment]
> I've hit a problem with the TryFrom changes that is unrelated to #3685, but that would probably also occur with this PR, related to the ambiguity of using `Self::Error` when `#[wasm_bindgen]` is applied to an enum containing an `Error` variant.
> 
> ```
> #[wasm_bindgen]
> pub enum LogLevel {
>     Trace,
>     Debug,
>     Info,
>     Warn,
>     Error,
> }
> ```
> 
> The way to solve that would be to use the fully qualified syntax for `Self`:
> 
> ```diff
> fn try_from_js_value(value: #wasm_bindgen::JsValue)
> -                   -> #wasm_bindgen::__rt::std::result::Result<Self, Self::Error> {
> +                   -> #wasm_bindgen::__rt::std::result::Result<Self, <Self as TryFromJsValue>::Error> {
> ```
> 
> I can open a PR myself after this one is merged if you prefer to merge this one as-is though.

I think this was already fixed in #3678, can you check if it still happens using a git version of the `wasm-bindgen` macro?

**Comment by dani-garcia** [comment]
My bad, seems this was fixed on master a bit ago, and I was looking at the struct changes of this PR instead of the enum changes. I just tested both master and this PR and they both work as expected, sorry for the confusion!

**Review by Liamolucko — APPROVED** [review]
I don't like regressing on #2231, but I guess the breaking changes are a higher priority.

**Comment by Liamolucko** [comment]
> > Alternatively we could make a `TryFromJsValue` derive macro, but that's a lot more work.
> 
> Even if we want to do that one day, it would still require `TryFromJsValue` trait to be public. I agree though this is weird, but I can't come up with a better solution.

By the way, what I meant by a `TryFromJsValue` derive macro was a derive macro named `TryFromJsValue` that implements `TryFrom<JsValue>`. I guess that is a bit confusing when there's also an internal trait with the same name, but the internal one could be renamed to `TryFromJsValueInternal` or something to avoid that (not that that needs to happen in this PR).

**Comment by daxpedda** [comment]
I'm gonna go ahead and merge this now and get the new release going.

**Pull request merged**
